File: live-test/go/normal.go
package main

import "fmt"

type Drawable interface {
	Draw() string
}

type Shape interface {
	Area() float64
	Describe() string
}

type baseShape struct{}

func (b baseShape) Describe() string { return "shape with unknown area" }

type Circle struct {
	baseShape
	Radius float64
}

func (c Circle) Area() float64 { return 3.14159 * c.Radius * c.Radius }
func (c Circle) Describe() string {
	return fmt.Sprintf("Circle with area %f", c.Area())
}
func (c Circle) Draw() string { return fmt.Sprintf("circle r=%f", c.Radius) }

type Rectangle struct {
	baseShape
	Width, Height float64
}

func (r Rectangle) Area() float64 { return r.Width * r.Height }
func (r Rectangle) Describe() string {
	return fmt.Sprintf("Rectangle with area %f", r.Area())
}
func (r Rectangle) Draw() string { return fmt.Sprintf("rect %fx%f", r.Width, r.Height) }

type Square struct{ Rectangle }

func NewSquare(side float64) Square {
	return Square{Rectangle: Rectangle{Width: side, Height: side}}
}

type ShapeCollection struct {
	Shapes []Shape
}

func (sc *ShapeCollection) Add(s Shape) {
	sc.Shapes = append(sc.Shapes, s)
}

func (sc *ShapeCollection) TotalArea() float64 {
	sum := 0.0
	for _, s := range sc.Shapes {
		sum += s.Area()
	}
	return sum
}

func (sc *ShapeCollection) Largest() Shape {
	var best Shape
	for _, s := range sc.Shapes {
		if best == nil || s.Area() > best.Area() {
			best = s
		}
	}
	return best
}

func MakeDefaultShapes() []Shape {
	return []Shape{
		Circle{Radius: 1.0},
		Rectangle{Width: 2.0, Height: 3.0},
		NewSquare(4.0),
	}
}

File: live-test/php/src/Normal.php
<?php
namespace Demo;

interface Drawable {
    public function draw(): string;
}

abstract class Shape {
    abstract public function area(): float;

    public function describe(): string {
        return get_class($this) . " with area " . $this->area();
    }
}

class Circle extends Shape implements Drawable {
    public function __construct(public readonly float $radius) {}

    public function area(): float {
        return 3.14159 * $this->radius * $this->radius;
    }

    public function draw(): string {
        return "circle r={$this->radius}";
    }
}

class Rectangle extends Shape implements Drawable {
    public function __construct(public readonly float $width, public readonly float $height) {}

    public function area(): float {
        return $this->width * $this->height;
    }

    public function draw(): string {
        return "rect {$this->width}x{$this->height}";
    }
}

class Square extends Rectangle {
    public function __construct(float $side) {
        parent::__construct($side, $side);
    }
}

class ShapeCollection {
    /** @var Shape[] */
    public array $shapes = [];

    public function add(Shape $shape): void {
        $this->shapes[] = $shape;
    }

    public function totalArea(): float {
        $sum = 0.0;
        foreach ($this->shapes as $s) {
            $sum += $s->area();
        }
        return $sum;
    }

    public function largest(): ?Shape {
        $best = null;
        foreach ($this->shapes as $s) {
            if ($best === null || $s->area() > $best->area()) {
                $best = $s;
            }
        }
        return $best;
    }
}

function makeDefaultShapes(): array {
    return [new Circle(1.0), new Rectangle(2.0, 3.0), new Square(4.0)];
}
